File: backend/src/app/routers/categories.py
from fastapi import APIRouter, status, HTTPException, Depends
from typing import List, Optional
from sqlmodel.ext.asyncio.session import AsyncSession
from pydantic import EmailStr

from models.models import Category
from schemas.schemas import CategoryCreate, CategoryResponse, CategoryUpdate, CategoryWithProducts
from crud.categories import create_category, read_category, read_categories, read_category_deleted, read_categories_deleted, update_category, delete_category, delete_categories, restaurar_category, restaurar_categories
from core.client import get_async_session
from core.security import get_current_user

#Router
router = APIRouter(prefix="/categories")

#EndPoints

#POST
@router.post("/create", status_code=status.HTTP_201_CREATED, response_model=CategoryResponse)
async def create(category: CategoryCreate, session: AsyncSession = Depends(get_async_session), current_user: EmailStr = Depends(get_current_user)) -> CategoryResponse:
    
    return await create_category(category=category, session=session)

#GET
@router.get("/read/{name}", status_code=status.HTTP_200_OK, response_model=CategoryResponse)
async def read(name: str, session: AsyncSession = Depends(get_async_session), current_user: EmailStr = Depends(get_current_user)) -> CategoryResponse:
    
    return await read_category(name=name, session=session)

@router.get("/read", status_code=status.HTTP_200_OK, response_model=List[CategoryResponse])
async def read2(session: AsyncSession = Depends(get_async_session), current_user: EmailStr = Depends(get_current_user)) -> List[CategoryResponse]:
    
    return await read_categories(session=session)

@router.get("/read-deleted/{name}", status_code=status.HTTP_200_OK, response_model=CategoryResponse)
async def read3(name: str, session: AsyncSession = Depends(get_async_session), current_user: EmailStr = Depends(get_current_user)) -> CategoryResponse:
    
    return await read_category_deleted(name=name, session=session)

@router.get("/read-deleted", status_code=status.HTTP_200_OK, response_model=List[CategoryResponse])
async def read4(session: AsyncSession = Depends(get_async_session), current_user: EmailStr = Depends(get_current_user)) -> List[CategoryResponse]:
    
    return await read_categories_deleted(session=session)

#PUT
@router.put("/update/{name}", status_code=status.HTTP_200_OK, response_model=CategoryResponse)
async def update(name: str, category: CategoryUpdate, session: AsyncSession = Depends(get_async_session), current_user: EmailStr = Depends(get_current_user)) -> CategoryResponse:
    
    return await update_category(name=name, category=category, session=session)

#DELETE
@router.delete("/delete/{name}", status_code=status.HTTP_200_OK, response_model=CategoryResponse)
async def delete(name: str, session: AsyncSession = Depends(get_async_session), current_user: EmailStr = Depends(get_current_user)) -> CategoryResponse:
    
    return await delete_category(name=name, session=session)

@router.delete("/delete", status_code=status.HTTP_200_OK, response_model=List[CategoryResponse])
async def delete2(session: AsyncSession = Depends(get_async_session), current_user: EmailStr = Depends(get_current_user)) -> List[CategoryResponse]:
    
    return await delete_categories(session=session)

#PATCH
@router.patch("/restaurar/{name}", status_code=status.HTTP_200_OK, response_model=CategoryResponse)
async def restaurar(name: str, session: AsyncSession = Depends(get_async_session), current_user: EmailStr = Depends(get_current_user)) -> CategoryResponse:
    
    return await restaurar_category(name=name, session=session)

@router.patch("/restaurar", status_code=status.HTTP_200_OK, response_model=List[CategoryResponse])
async def restaurar2(session: AsyncSession = Depends(get_async_session), current_user: EmailStr = Depends(get_current_user)) -> List[CategoryResponse]:
    
    return await restaurar_categories(session=session)

File: frontend/src/features/admin/api/categories.ts
import { client } from "@/lib/api/client.gen";
import type {
  CategoryCreate,
  CategoryResponse,
  CategoryUpdate,
} from "@/lib/api/types.gen";

export async function fetchAdminCategories(): Promise<CategoryResponse[]> {
  const response = await client.get({ url: "/categories/read" });
  if (response.error) {
    throw new Error("Error al cargar las categorías");
  }
  return (response.data as CategoryResponse[]) ?? [];
}

export async function fetchAdminDeletedCategories(): Promise<CategoryResponse[]> {
  const response = await client.get({ url: "/categories/read-deleted" });
  if (response.error) {
    throw new Error("Error al cargar las categorías eliminadas");
  }
  return (response.data as CategoryResponse[]) ?? [];
}

export async function createCategory(
  data: CategoryCreate
): Promise<CategoryResponse> {
  const response = await client.post({
    url: "/categories/create",
    body: data,
  });
  if (response.error || !response.data) {
    throw new Error("Error al crear la categoría");
  }
  return response.data as CategoryResponse;
}

export async function updateCategory(
  name: string,
  data: CategoryUpdate
): Promise<CategoryResponse> {
  const response = await client.put({
    url: "/categories/update/{name}",
    path: { name },
    body: data,
  });
  if (response.error || !response.data) {
    throw new Error("Error al actualizar la categoría");
  }
  return response.data as CategoryResponse;
}

export async function deleteCategory(name: string): Promise<void> {
  const response = await client.delete({
    url: "/categories/delete/{name}",
    path: { name },
  });
  if (response.error) {
    throw new Error("Error al eliminar la categoría");
  }
}

export async function restoreCategory(name: string): Promise<CategoryResponse> {
  const response = await client.patch({
    url: "/categories/restaurar/{name}",
    path: { name },
  });
  if (response.error || !response.data) {
    throw new Error("Error al restaurar la categoría");
  }
  return response.data as CategoryResponse;
}
